File: architecture-skills/design-patterns/examples/chain-of-responsibility/chain-of-responsibility.swift
// Chain of Responsibility pattern in Swift.
//
// Support ticket escalation: a protocol-based chain where each level
// resolves tickets up to its competence or forwards them. UIKit's
// responder chain is the canonical platform incarnation.

struct Ticket {
    let subject: String
    let severity: Int // 1 (trivial) .. 4 (critical)
}

protocol Handler: AnyObject {
    var next: Handler? { get set }
    func handle(_ ticket: Ticket)
}

extension Handler {
    @discardableResult
    func setNext(_ handler: Handler) -> Handler {
        next = handler
        return handler // enables a.setNext(b).setNext(c)
    }

    func forward(_ ticket: Ticket) {
        if let next {
            next.handle(ticket)
        } else {
            print("UNRESOLVED: \"\(ticket.subject)\" (severity \(ticket.severity))"
                + " -> escalate to engineering")
        }
    }
}

final class SupportLevel: Handler {
    var next: Handler?
    private let name: String
    private let maxSeverity: Int

    init(name: String, maxSeverity: Int) {
        self.name = name
        self.maxSeverity = maxSeverity
    }

    func handle(_ ticket: Ticket) {
        if ticket.severity <= maxSeverity {
            print("\(name) resolved \"\(ticket.subject)\" (severity \(ticket.severity))")
        } else {
            forward(ticket)
        }
    }
}

let chain = SupportLevel(name: "Help bot", maxSeverity: 1)
chain.setNext(SupportLevel(name: "Frontline agent", maxSeverity: 2))
    .setNext(SupportLevel(name: "Support manager", maxSeverity: 3))

let tickets = [
    Ticket(subject: "Reset my password", severity: 1),
    Ticket(subject: "Invoice shows wrong amount", severity: 2),
    Ticket(subject: "Account data corrupted", severity: 3),
    Ticket(subject: "Full region outage", severity: 4),
]

for ticket in tickets {
    chain.handle(ticket)
}

File: architecture-skills/design-patterns/examples/chain-of-responsibility/chain-of-responsibility.java
// Chain of Responsibility pattern in Java.
//
// Support ticket escalation: an abstract handler holds the next link;
// each level resolves tickets up to its competence or forwards them.
// Servlet filters are the same pattern built into the platform.
//
// The top-level class is package-private so the filename does not have
// to match a (hyphen-containing) public class name.

import java.util.List;

class ChainOfResponsibilityDemo {

    record Ticket(String subject, int severity) {} // severity 1..4

    abstract static class Handler {
        private Handler next;

        Handler setNext(Handler handler) {
            this.next = handler;
            return handler; // enables a.setNext(b).setNext(c)
        }

        void handle(Ticket ticket) {
            if (next != null) {
                next.handle(ticket);
            } else {
                System.out.printf("UNRESOLVED: \"%s\" (severity %d) -> escalate to engineering%n",
                        ticket.subject(), ticket.severity());
            }
        }
    }

    static final class SupportLevel extends Handler {
        private final String name;
        private final int maxSeverity;

        SupportLevel(String name, int maxSeverity) {
            this.name = name;
            this.maxSeverity = maxSeverity;
        }

        @Override
        void handle(Ticket ticket) {
            if (ticket.severity() <= maxSeverity) {
                System.out.printf("%s resolved \"%s\" (severity %d)%n",
                        name, ticket.subject(), ticket.severity());
            } else {
                super.handle(ticket);
            }
        }
    }

    public static void main(String[] args) {
        Handler chain = new SupportLevel("Help bot", 1);
        chain.setNext(new SupportLevel("Frontline agent", 2))
                .setNext(new SupportLevel("Support manager", 3));

        List<Ticket> tickets = List.of(
                new Ticket("Reset my password", 1),
                new Ticket("Invoice shows wrong amount", 2),
                new Ticket("Account data corrupted", 3),
                new Ticket("Full region outage", 4));

        for (Ticket ticket : tickets) {
            chain.handle(ticket);
        }
    }
}
